### ISSUE TITLE
Allow enforcing single/double-quotes without preventing use of template literals

### ISSUE THREAD

**Issue by sindresorhus** [eslint/eslint#5234]
How it worked in ESLint 1.0.0 worked perfectly. I could enforce single-quotes everywhere `'quotes': [2, 'single']` and still use template literals when convenient, e.g. if the content had single-quotes so not having to escape it. With ESLint 2.0.0 template literals are now awkwardly forced into the rule and there's no way to enforce single-quotes without also preventing use of template literals when they're not strictly required. I don't want to use double-quotes to prevent having to escape strings containing single-quotes. I want to use template literals, or rather, I already do, and this change makes everything break for me...

I've read both https://github.com/eslint/eslint/issues/5147 and https://github.com/eslint/eslint/issues/3090, but I disagree with the conclusion and I think they're missing this use-case.

To be clear, this used to pass:

```
~/dev
❯ eslint --version                                                                           
v2.0.0
                                                                                                                                               
~/dev
❯ echo "var foo = \`bar 'foo' baz\`;" | eslint --stdin --env=es6 --rule='quotes: [2, single]'

<text>
  1:11  error  Strings must use singlequote  quotes

✖ 1 problem (1 error, 0 warnings)
```

I would like an option to allow this functionality.

**Comment by guyellis** [comment]
:+1: I've just run into this in upgrading to ESLint 2.0.0 as well. Just to be sure that I understand what @sindresorhus is asking, this is what I'd like see:

`"quotes": [2, "backtick", "single"]`

i.e. Allow both backticks and singles but not doubles.

This obviously opens up the discussion/ability to allow any permutation of the 3 quote types in JS.

**Comment by BYK** [comment]
I think this would be covered with the second "avoid-escape" option if it is not already.

**Comment by BYK** [comment]
>if it is not already.

And it seems like it is not.

**Comment by Qix-** [comment]
:+1: yeah this was a bad move. I dont use template strings everywhere - only when they're convenient.

**Comment by nzakas** [comment]
I like @guyellis's suggestion. Anyone want to submit a PR?

**Comment by BYK** [comment]
@nzakas I'd love to but wouldn't that suggestion collide with the `avoid-escape` option?

**Comment by nzakas** [comment]
I don't think so. There's no limit to the number of items in the array. We'd just need to ensure that we don't assume any specific position for each option.

**Comment by IanVS** [comment]
I believe the way schemas work is they do assume a position for each option, don't they?  Instead of potential collisions, I would suggest changing the first option from `enum` to `oneOf: [ enum, array ]`.  So, the final config might look like:

```json
quotes": [2, ["backtick", "single"]]
```

Explicitly placing them in an array is a bit more clear to me in any case, and I think it will avoid problems with schema validation and autoconfig.

**Comment by BYK** [comment]
I think `avoid-escape` option should cover this. Currently it allows `"Fred's"` but it doesn't allow `` `Fred's` `` with the config `quotes": [2, "single", "avoid-escape"]`

**Comment by IanVS** [comment]
We've said previously that backticks should only be allowed if they are providing functionality which is impossible with the chosen quote type.  If you want to avoid escaping a single quote, you can use a double quote, right?

**Comment by nzakas** [comment]
@IanVS I don't think embedding another array makes sense as an end user. We have other rules that have multiple items, I'm not sure why this one would be a problem.

@byk I don't think that's correct. `avoid-escape ` lets you switch quote marks only if you'd have to escape the preferred quote Mark, which is not the issue here.

**Comment by vdemedes** [comment]
Agree with @sindresorhus, this change brings unexpected behavior. Would be great to keep the old one.

**Comment by Qix-** [comment]
> I don't think embedding another array makes sense as an end user.

Depends on the user I suppose.

```json
"quotes": [2, "backtick", "single"]
```

is super clean, but doesn't make sense on the technical (schema) level.

```json
"quotes": [2, ["backtick", "single"]]
```

makes sense on the technical level, and to technical people, but not to those who prefer aesthetics to code.

Are there any other rules that have embedded arrays/multiple choices? Or is this the first?

**Comment by SpadeAceman** [comment]
Since [backticks aren't purely equivalent to single or double quotes](https://hacks.mozilla.org/2015/05/es6-in-depth-template-strings-2/), ideally another option is needed beyond `avoid-escape`, to ensure that the appropriate quote type is employed for the functionality being used (e.g., single quotes for regular strings, backticks for template literals/multi-line strings). `prefer-template` doesn't really cover this, checking only non-backtick strings for concatenation.

**Comment by btmills** [comment]
We can ignore the schema here - figure out what the right options are, and then we (I volunteer if necessary :hand:) can figure out a schema to make it work.

**Comment by michaelficarra** [comment]
The `avoid-escape` option is not sufficient here. With config `"quotes": [2, "single", "avoid-escape"]`, I still want to be warned if I use an untagged single-line template with no interpolation. I am expecting the `avoid-escape` option here to allow me to use double quotes, not to allow the use of untagged single-line templates with no interpolation. It seems the rule requested by #3090 really should not have been included as part of the `quotes` rule, and should have been its own rule, as originally requested.

**Comment by feross** [comment]
+1 to splitting out template-strings-only-as-needed from the quotes rule.

While we're considering changes to the `quotes` rule, I'll mention that `standard` would like to enforce this:

1. default rule: use single quotes
2. if string contains a single quote: use single quotes (w/ escaping), or double quotes (to avoid escaping)
3. if string contains single **and** double quotes: use single quotes (w/ escaping), or backticks (to avoid escaping)

Right now, there's no way to allow the third case. That issue came up here: https://github.com/feross/standard/issues/421

This could maybe be added as an additional exception to the `"avoid-escape"` case.

**Comment by michaelficarra** [comment]
@feross That is a very complex style preference. I don't think something like that is appropriate for a core rule.

**Comment by nzakas** [comment]
Agreed. @feross, that's far too specific for a core rule, and if you could please not add enhancement requests to already-accepted issues, I'd appreciate it. It gets very hard to track conversations and requests when that happens.

**Comment by feross** [comment]
@michaelficarra @nzakas I should have been clearer. My request is related to the current discussion.

I think we can satisfy everyone if we:

1) Add new rule to disallow untagged single-line template with no interpolation. (e.g. `"use-template-string-only-when-necessary"`)

2) Change `"avoid-escape"` to allow using any quote type to avoid escaping, including backticks. This is the pre-v2 behavior of the option.

Then users who want the behavior @sindresorhus describes can just use `"avoid-escape"`.

**Comment by feross** [comment]
> @IanVS We've said previously that backticks should only be allowed if they are providing functionality which is impossible with the chosen quote type. If you want to avoid escaping a single quote, you can use a double quote, right?

Yes, but what about the case where the user has a string that contains a single quote and a double quote? Then using a template string is necessary to avoid escaping.

If the changes I suggested in my previous comment are made, and I enable both rules, i.e.:

```js
{
  "quotes": [2, "single", "avoid-escape"], // pre v2 behavior for "avoid-escape"
  "use-template-string-only-when-necessary": 2
}
```

Then I would want the following code to be allowed:

```js
var str = `John said "You're such a good 'teacher' sometimes."`
```

Because I'm using backticks to avoid escaping, `"avoid-escape"` won't complain. And because the template string is necessary here, the second rule won't complain either.

**Comment by michaelficarra** [comment]
@feross The template is *not* necessary there. It is untagged, it has no interpolation, and it is not multiline. The `use-template-string-only-when-necessary` rule would report that.

**Comment by ksmithut** [comment]
I'm not sure if this is related, but I believe it to be a valid use case that should be included somewhere in eslint.

```
return `The string has multiple lines, but the first line is too long ` +
  `to include on a single line (with max line length rules). Somewhere ` +
  `in this string there is use of interpolation ${someVar}, but we want ` +
  `to maintain consistent quotes.`
```

The reason we wouldn't want to use the multiline feature of template strings in some cases is if this string gets returned in an api response or log where we don't actually want to log the newline and extra space characters.

**Comment by jasonbarone** [comment]
I can't seem to find a solution after upgrading to 2.0 and I feel like @feross' comment above would be a good solution. My use case is very similar. Am I missing something on ESlint, is there no other way around this aside from removing quote style enforcement? To avoid dealing with this I've been reverting back to ESlint 1.0.

```
{
  "quotes": [2, "double" ],
  "use-template-string-only-when-necessary": 2
}
```
Would allow the following;
```
const hook = `<div class="hook__placeholder"></div>`;
```

**Comment by IanVS** [comment]
@nzakas what is the accepted change here?  Do we want to break this off into a new rule as has been suggested?

**Comment by MiguelCastillo** [comment]
Opinionated vote is to have separate rules.  It is more clear to me.  In projects I am involved in they are both used and they certainly don't mean the same thing.  A tick isn't a quote...  Although a tick does encompass the functionality of a quote.  `template-string` seems very appropriate.  And perhaps even add some of the semantics provided with `avoid-escape` if we feel that we really need it.

I don't see ticks in the wild very frequently being used in import statements, even though with the restrictions we are imposing we have to:

```
import x from `x`;
```

One thing that concerns me is that the mere fact that we have to rephrase several times the intricate differences when ticks are valid in the context of a quote tells me that we are adding complexity in just explaining to the user how to configure eslint. I am not taking into account development complexities.

Also, keeping quotes as is and adding a separate rule for managing ticks gives us breathing room and flexibility for enhancing their behavior without needing to work through the intricate differences each time. Perhaps we want to add a future restriction that template tags can only be of a certain color...

The unfortunate part is that as of right now, my team isn't able to upgrade (stuck in 1.x) because we have a need to use both, single quotes and ticks.  :-/

**Comment by nzakas** [comment]
@IanVS "accepted" was in response to https://github.com/eslint/eslint/issues/5234#issuecomment-183676342

I still think just having an option of `quotes` that says "allow template literals in addition to my preference" is the simplest and easiest way to resolve this. I don't see a lot of value in splitting out an entire separate rule for what amounts to one exception. 

Maybe if we clean it up a bit, we can have a new schema that's a bit more readable, like:

```
quotes: ["error", "double", {
    avoidEscape: true,
    allowTemplateLiterals: true
}]
```

The `quotes` rule is one of the oldest rules we have, so it does have a bit of the legacy option feel to it. This would clean it up nicely (I think).

**Comment by BarryThePenguin** [comment]
Would you look at a PR deprecating
```
quotes: ["error", "double", "avoid-escape"]
```

and using the suggested?

```
quotes: ["error", "double", {
    avoidEscape: true,
    allowTemplateLiterals: true
}]
```

**Comment by ilyavolodin** [comment]
@BarryThePenguin We can't deprecate old style, at least not yet. That would be a breaking change, and we can only do those during major releases. I think we should just modify the rule to take both old and new style of options. We have a few rules that do that already.

**Comment by BarryThePenguin** [comment]
Sure. I probably should have said 'mark as deprecated'. I've had a look into this have a solution for maintaining the old and new style of options. Then in the next major, remove the old style.

**Comment by kaicataldo** [comment]
Just to be clear, https://github.com/eslint/eslint/issues/5234#issuecomment-196530074 is the agreed upon configuration we're shooting for, right? If so, I'll take a look at this.

**Comment by BarryThePenguin** [comment]
@kaicataldo I already have a solution, but haven't created a PR as I wasn't sure if it was agreed or not...

**Comment by kaicataldo** [comment]
@BarryThePenguin Awesome, definitely don't want to duplicate efforts :)

**Comment by IanVS** [comment]
@BarryThePenguin Feel free to submit a PR, the issue is accepted.

**Comment by bobsilverberg** [comment]
@BarryThePenguin Just wondering if you are planning on submitting a PR, and if so what your timing is. We just updated eslint and are getting warnings which the fix to this issue would eliminate. I'd like to hold off updating my code to address the warnings if a fix is coming soon.

**Comment by BarryThePenguin** [comment]
@bobsilverberg PR is linked above

**Comment by bobsilverberg** [comment]
Thanks @BarryThePenguin! I look forward to seeing that land. :)

**Comment by romanzoller** [comment]
If I understand correctly, most people here agreed that a separate `use-template-string-only-when-necessary` rule would be a good idea, but as far as I can tell, it has never been implemented. Should I create a new issue for it?

**Comment by not-an-aardvark** [comment]
That behavior is already part of the `quotes` rule.

**Comment by romanzoller** [comment]
As far as I understand, not really, or not anymore, the doc for the `quotes` rule used to say

> When using `"single"` or `"double"`, template literals that don't contain a substitution, don't contain a line break and aren't tagged templates, are flagged as problems, even with the `"avoid-escape"` option.

And then fb3c2eb7ce6a07507e48b0fe1986dc15ebc847b2 added:

> However they are not problems when `"allowTemplateLiterals"` is used.

which was later removed altogether.

I thought the idea of `use-template-string-only-when-necessary` was that template literals are allowed, but only if they cannot be replaced by single or double quotes. This is the use case outlined in #3090 which was also mentioned here.
